From linux-kernel@vger.kernel.org Tue Jan 13 00:59:23 2009
X-Envelope-From: <git-commits-head-owner@vger.kernel.org>
X-Envelope-To: <vger@sl-h.de>
X-Delivery-Time: 1231804769
X-UID: 403526
Return-Path: <git-commits-head-owner@vger.kernel.org>
X-RZG-CLASS-ID: mi
Received: from vger.kernel.org ([209.132.176.167])
	by mailin.webmailer.de (zeb mi54) (RZmta 18.7)
	with ESMTP id K01448l0CNaQNh for <vger@sl-h.de>;
	Tue, 13 Jan 2009 00:59:29 +0100 (MET)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
	id S1755018AbZALX71 (ORCPT <rfc822;vger@sl-h.de>);
	Mon, 12 Jan 2009 18:59:27 -0500
Received: (majordomo@vger.kernel.org) by vger.kernel.org id S1754994AbZALX71
	(ORCPT <rfc822;git-commits-head-outgoing>);
	Mon, 12 Jan 2009 18:59:27 -0500
Received: from hera.kernel.org ([140.211.167.34]:33157 "EHLO hera.kernel.org"
	rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
	id S1754799AbZALX70 (ORCPT
	<rfc822;git-commits-head@vger.kernel.org>);
	Mon, 12 Jan 2009 18:59:26 -0500
Received: from hera.kernel.org (IDENT:U2FsdGVkX1+UCZ49qS+I1DtI9jm82S+eAP1Ur3oGG9k@localhost [127.0.0.1])
	by hera.kernel.org (8.14.2/8.14.2) with ESMTP id n0CNxO0R015835
	(version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-SHA bits=256 verify=NO)
	for <git-commits-head@vger.kernel.org>; Mon, 12 Jan 2009 23:59:24 GMT
Received: (from dwmw2@localhost)
	by hera.kernel.org (8.14.2/8.13.1/Submit) id n0CNxNiL015817
	for git-commits-head@vger.kernel.org; Mon, 12 Jan 2009 23:59:23 GMT
Date:	Mon, 12 Jan 2009 23:59:23 GMT
Message-Id: <200901122359.n0CNxNiL015817@hera.kernel.org>
From:	Linux Kernel Mailing List <linux-kernel@vger.kernel.org>
To: git-commits-head@vger.kernel.org
Subject: ALSA: caiaq - Fix Oops with MIDI
MIME-Version: 1.0
Content-Type: text/plain;
  charset=UTF-8
Content-Transfer-Encoding: 8bit
X-Git-Commit: f3f80a9205da74fa56d613f4c14b88b6e4e6caa8
X-Git-Parent: 9e42d0cf5020aaf217433cad1a224745241d212a
X-Virus-Scanned: ClamAV 0.93.3/8856/Mon Jan 12 16:36:19 2009 on hera.kernel.org
X-Virus-Status:	Clean
X-Spam-Status: No, score=-2.6 required=5.0 tests=AWL,BAYES_00,
	UNPARSEABLE_RELAY autolearn=ham version=3.2.5
X-Spam-Checker-Version:	SpamAssassin 3.2.5 (2008-06-10) on hera.kernel.org
X-Greylist: Sender IP whitelisted, not delayed by milter-greylist-4.0 (hera.kernel.org [127.0.0.1]); Mon, 12 Jan 2009 23:59:26 +0000 (UTC)
Sender:	git-commits-head-owner@vger.kernel.org
Precedence: bulk
List-ID: <git-commits-head.vger.kernel.org>
X-Mailing-List:	git-commits-head@vger.kernel.org
Status: R
X-Status: NC
X-KMail-EncryptionState:  
X-KMail-SignatureState:  
X-KMail-MDN-Sent:  

Gitweb:     http://git.kernel.org/git/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commit;h=f3f80a9205da74fa56d613f4c14b88b6e4e6caa8
Commit:     f3f80a9205da74fa56d613f4c14b88b6e4e6caa8
Parent:     9e42d0cf5020aaf217433cad1a224745241d212a
Author:     Takashi Iwai <tiwai@suse.de>
AuthorDate: Thu Jan 8 15:32:56 2009 +0100
Committer:  Takashi Iwai <tiwai@suse.de>
CommitDate: Thu Jan 8 15:32:56 2009 +0100

    ALSA: caiaq - Fix Oops with MIDI
    
    The snd-usb-caiaq driver causes Oops occasionally when accessing MIDI
    devices.  This patch fixes the Oops and invalid URB submission errors
    as well.
    
    Cc: stable@kernel.org
    Signed-off-by: Takashi Iwai <tiwai@suse.de>
---
 sound/usb/caiaq/caiaq-device.h |    1 +
 sound/usb/caiaq/caiaq-midi.c   |   32 ++++++++++++++++++--------------
 2 files changed, 19 insertions(+), 14 deletions(-)

diff --git a/sound/usb/caiaq/caiaq-device.h b/sound/usb/caiaq/caiaq-device.h
index f9fbdba..ab56e73 100644
--- a/sound/usb/caiaq/caiaq-device.h
+++ b/sound/usb/caiaq/caiaq-device.h
@@ -75,6 +75,7 @@ struct snd_usb_caiaqdev {
 	wait_queue_head_t ep1_wait_queue;
 	wait_queue_head_t prepare_wait_queue;
 	int spec_received, audio_parm_answer;
+	int midi_out_active;
 
 	char vendor_name[CAIAQ_USB_STR_LEN];
 	char product_name[CAIAQ_USB_STR_LEN];
diff --git a/sound/usb/caiaq/caiaq-midi.c b/sound/usb/caiaq/caiaq-midi.c
index 30b57f9..f19fd36 100644
--- a/sound/usb/caiaq/caiaq-midi.c
+++ b/sound/usb/caiaq/caiaq-midi.c
@@ -59,6 +59,11 @@ static int snd_usb_caiaq_midi_output_open(struct snd_rawmidi_substream *substrea
 
 static int snd_usb_caiaq_midi_output_close(struct snd_rawmidi_substream *substream)
 {
+	struct snd_usb_caiaqdev *dev = substream->rmidi->private_data;
+	if (dev->midi_out_active) {
+		usb_kill_urb(&dev->midi_out_urb);
+		dev->midi_out_active = 0;
+	}
 	return 0;
 }
 
@@ -69,7 +74,8 @@ static void snd_usb_caiaq_midi_send(struct snd_usb_caiaqdev *dev,
 	
 	dev->midi_out_buf[0] = EP1_CMD_MIDI_WRITE;
 	dev->midi_out_buf[1] = 0; /* port */
-	len = snd_rawmidi_transmit_peek(substream, dev->midi_out_buf+3, EP1_BUFSIZE-3);
+	len = snd_rawmidi_transmit(substream, dev->midi_out_buf + 3,
+				   EP1_BUFSIZE - 3);
 	
 	if (len <= 0)
 		return;
@@ -79,24 +85,24 @@ static void snd_usb_caiaq_midi_send(struct snd_usb_caiaqdev *dev,
 	
 	ret = usb_submit_urb(&dev->midi_out_urb, GFP_ATOMIC);
 	if (ret < 0)
-		log("snd_usb_caiaq_midi_send(%p): usb_submit_urb() failed, %d\n",
-				substream, ret);
+		log("snd_usb_caiaq_midi_send(%p): usb_submit_urb() failed,"
+		    "ret=%d, len=%d\n",
+		    substream, ret, len);
+	else
+		dev->midi_out_active = 1;
 }
 
 static void snd_usb_caiaq_midi_output_trigger(struct snd_rawmidi_substream *substream, int up)
 {
 	struct snd_usb_caiaqdev *dev = substream->rmidi->private_data;
 	
-	if (dev->midi_out_substream != NULL)
-		return;
-	
-	if (!up) {
+	if (up) {
+		dev->midi_out_substream = substream;
+		if (!dev->midi_out_active)
+			snd_usb_caiaq_midi_send(dev, substream);
+	} else {
 		dev->midi_out_substream = NULL;
-		return;
 	}
-	
-	dev->midi_out_substream = substream;
-	snd_usb_caiaq_midi_send(dev, substream);
 }
 
 
@@ -161,16 +167,14 @@ int snd_usb_caiaq_midi_init(struct snd_usb_caiaqdev *device)
 void snd_usb_caiaq_midi_output_done(struct urb* urb)
 {
 	struct snd_usb_caiaqdev *dev = urb->context;
-      	char *buf = urb->transfer_buffer;
 	
+	dev->midi_out_active = 0;
 	if (urb->status != 0)
 		return;
 
 	if (!dev->midi_out_substream)
 		return;
 
-	snd_rawmidi_transmit_ack(dev->midi_out_substream, buf[2]);
-	dev->midi_out_substream = NULL;
 	snd_usb_caiaq_midi_send(dev, dev->midi_out_substream);
 }
 
--
To unsubscribe from this list: send the line "unsubscribe git-commits-head" in
the body of a message to majordomo@vger.kernel.org
More majordomo info at  http://vger.kernel.org/majordomo-info.html

